File: src/notes.ts
import { mkdir, readdir, readFile, rm, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { isValidYMD } from "./models.ts";
import type { DataPaths } from "./paths.ts";

export const NOTE_TYPES = ["subjective", "observation", "lesson"] as const;
export type NoteType = (typeof NOTE_TYPES)[number];

export const NOTE_AUTHORS = ["athlete", "coach"] as const;
export type NoteAuthor = (typeof NOTE_AUTHORS)[number];

export interface NoteRecord {
  id: string;
  date: string;
  type: NoteType;
  workout_id?: number;
  tags?: string[];
  body: string;
  author: NoteAuthor;
}

const CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

export function ulid(now: Date): string {
  let ms = now.getTime();
  let time = "";
  for (let i = 0; i < 10; i++) {
    time = CROCKFORD[ms % 32] + time;
    ms = Math.floor(ms / 32);
  }
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  let rand = "";
  for (const b of bytes) {
    rand += CROCKFORD[b % 32];
  }
  return time + rand;
}

export function localISO(d: Date): string {
  const pad = (n: number, width = 2) => String(Math.abs(n)).padStart(width, "0");
  const offsetMinutes = -d.getTimezoneOffset();
  const sign = offsetMinutes >= 0 ? "+" : "-";
  const offset = `${sign}${pad(Math.floor(Math.abs(offsetMinutes) / 60))}:${pad(Math.abs(offsetMinutes) % 60)}`;
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}T${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}${offset}`;
}

export function serializeNote(note: NoteRecord): string {
  const ordered: Record<string, unknown> = { id: note.id, date: note.date, type: note.type };
  if (note.workout_id != null) ordered.workout_id = note.workout_id;
  if (note.tags != null && note.tags.length > 0) ordered.tags = note.tags;
  ordered.body = note.body;
  ordered.author = note.author;
  return JSON.stringify(ordered);
}

export function isNoteShaped(parsed: unknown): parsed is NoteRecord {
  const note = parsed as NoteRecord;
  return (
    typeof note?.id === "string" &&
    typeof note?.date === "string" &&
    typeof note?.body === "string" &&
    (NOTE_TYPES as readonly string[]).includes(note?.type) &&
    (NOTE_AUTHORS as readonly string[]).includes(note?.author) &&
    /^\d{4}-\d{2}-\d{2}T.*(?:Z|[+-]\d{2}:\d{2})$/.test(note.date) &&
    isValidYMD(note.date.slice(0, 10)) &&
    !Number.isNaN(new Date(note.date).getTime()) &&
    (note.workout_id === undefined ||
      (typeof note.workout_id === "number" && Number.isFinite(note.workout_id))) &&
    (note.tags === undefined ||
      (Array.isArray(note.tags) && note.tags.every((t) => typeof t === "string")))
  );
}

function parseNote(raw: string): NoteRecord | null {
  try {
    const parsed = JSON.parse(raw) as unknown;
    return isNoteShaped(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

function compareNotes(a: NoteRecord, b: NoteRecord): number {
  const aMs = new Date(a.date).getTime();
  const bMs = new Date(b.date).getTime();
  if (aMs !== bMs) return aMs < bMs ? -1 : 1;
  if (a.id !== b.id) return a.id < b.id ? -1 : 1;
  return 0;
}

export async function writeNote(paths: DataPaths, note: NoteRecord): Promise<void> {
  await mkdir(paths.notesDir, { recursive: true });
  await writeFile(join(paths.notesDir, `${note.id}.json`), `${serializeNote(note)}\n`, "utf-8");
}

interface LooseEntry {
  note: NoteRecord;
  file: string;
}

async function readLooseEntries(paths: DataPaths): Promise<LooseEntry[]> {
  const entries: LooseEntry[] = [];
  let files: string[];
  try {
    files = await readdir(paths.notesDir);
  } catch {
    return entries;
  }
  for (const f of files.sort()) {
    if (!f.endsWith(".json")) continue;
    try {
      const note = parseNote(await readFile(join(paths.notesDir, f), "utf-8"));
      if (note != null) entries.push({ note, file: f });
    } catch {}
  }
  return entries;
}

async function readLooseNotes(paths: DataPaths): Promise<Map<string, NoteRecord>> {
  const notes = new Map<string, NoteRecord>();
  for (const entry of await readLooseEntries(paths)) {
    notes.set(entry.note.id, entry.note);
  }
  return notes;
}

async function readArchivedNotes(paths: DataPaths): Promise<Map<string, NoteRecord>> {
  const notes = new Map<string, NoteRecord>();
  let files: string[];
  try {
    files = await readdir(paths.archiveDir);
  } catch {
    return notes;
  }
  for (const f of files) {
    if (!f.endsWith(".jsonl")) continue;
    let text: string;
    try {
      text = await readFile(join(paths.archiveDir, f), "utf-8");
    } catch {
      continue;
    }
    for (const line of text.split("\n")) {
      if (line.trim() === "") continue;
      const note = parseNote(line);
      if (note != null && !notes.has(note.id)) notes.set(note.id, note);
    }
  }
  return notes;
}

export async function readAllNotes(paths: DataPaths): Promise<NoteRecord[]> {
  const archived = await readArchivedNotes(paths);
  const loose = await readLooseNotes(paths);
  for (const [id, note] of loose) {
    archived.set(id, note);
  }
  return [...archived.values()].sort(compareNotes);
}

export interface NoteFilter {
  type?: string;
  since?: string;
  workoutId?: number;
}

export function filterNotes(notes: NoteRecord[], filter: NoteFilter): NoteRecord[] {
  return notes.filter((n) => {
    if (filter.type && n.type !== filter.type) return false;
    if (filter.since && n.date.slice(0, 10) < filter.since) return false;
    if (filter.workoutId != null && n.workout_id !== filter.workoutId) return false;
    return true;
  });
}

const COMPACT_AGE_DAYS = 7;

async function readArchiveYear(
  paths: DataPaths,
  year: number,
): Promise<{ notes: NoteRecord[]; safeToRewrite: boolean }> {
  const notes: NoteRecord[] = [];
  let text: string;
  try {
    text = await readFile(paths.archiveFile(year), "utf-8");
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    return { notes, safeToRewrite: code === "ENOENT" };
  }
  const seen = new Set<string>();
  for (const line of text.split("\n")) {
    if (line.trim() === "") continue;
    const note = parseNote(line);
    if (note == null) return { notes, safeToRewrite: false };
    if (seen.has(note.id)) return { notes, safeToRewrite: false };
    seen.add(note.id);
    notes.push(note);
  }
  return { notes, safeToRewrite: true };
}

export interface CompactResult {
  archived: number;
  years: number[];
  skippedYears: number[];
}

export async function compactNotes(paths: DataPaths, now: Date): Promise<CompactResult> {
  const cutoff = new Date(now);
  cutoff.setDate(cutoff.getDate() - COMPACT_AGE_DAYS);
  const cutoffMs = cutoff.getTime();

  const entries = await readLooseEntries(paths);
  const deduped = new Map<string, NoteRecord>();
  const contentById = new Map<string, string>();
  const divergent = new Set<string>();
  for (const e of entries) {
    const content = serializeNote(e.note);
    const prior = contentById.get(e.note.id);
    if (prior != null && prior !== content) divergent.add(e.note.id);
    contentById.set(e.note.id, content);
    deduped.set(e.note.id, e.note);
  }
  const eligible = [...deduped.values()].filter(
    (n) => !divergent.has(n.id) && new Date(n.date).getTime() < cutoffMs,
  );
  if (eligible.length === 0) return { archived: 0, years: [], skippedYears: [] };

  const byYear = new Map<number, NoteRecord[]>();
  for (const note of eligible) {
    const year = Number(note.date.slice(0, 4));
    if (!byYear.has(year)) byYear.set(year, []);
    byYear.get(year)!.push(note);
  }

  await mkdir(paths.archiveDir, { recursive: true });
  let archivedCount = 0;
  const years: number[] = [];
  const skippedYears: number[] = [];
  for (const [year, notes] of byYear) {
    const existing = await readArchiveYear(paths, year);
    if (!existing.safeToRewrite) {
      skippedYears.push(year);
      continue;
    }
    const merged = new Map<string, NoteRecord>();
    for (const n of existing.notes) merged.set(n.id, n);
    for (const n of notes) merged.set(n.id, n);
    const lines = [...merged.values()].sort(compareNotes).map(serializeNote);
    try {
      await writeFile(paths.archiveFile(year), `${lines.join("\n")}\n`, "utf-8");
    } catch {
      skippedYears.push(year);
      continue;
    }
    const archivedIds = new Set(notes.map((n) => n.id));
    for (const e of entries) {
      if (archivedIds.has(e.note.id)) {
        try {
          await rm(join(paths.notesDir, e.file), { force: true });
        } catch {}
      }
    }
    archivedCount += notes.length;
    years.push(year);
  }

  return { archived: archivedCount, years: years.sort(), skippedYears: skippedYears.sort() };
}

File: internal/notes/notes.go
package notes

import (
	"bytes"
	"crypto/rand"
	"encoding/base32"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/richhaase/c2/internal/atomicfile"
	"github.com/richhaase/c2/internal/jsonx"
	"github.com/richhaase/c2/internal/models"
	"github.com/richhaase/c2/internal/paths"
)

var Types = []string{"subjective", "observation", "lesson"}

var Authors = []string{"athlete", "coach"}

const compactAgeDays = 7

const crockford = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"

var crockfordEncoding = base32.NewEncoding(crockford).WithPadding(base32.NoPadding)

type Record struct {
	ID        string   `json:"id"`
	Date      string   `json:"date"`
	Type      string   `json:"type"`
	WorkoutID *int64   `json:"workout_id,omitempty"`
	Tags      []string `json:"tags,omitempty"`
	Body      string   `json:"body"`
	Author    string   `json:"author"`
}

func ULID(now time.Time) (string, error) {
	ms := now.UnixMilli()
	if ms < 0 || ms >= 1<<48 {
		return "", fmt.Errorf("note timestamp is outside the ULID range")
	}
	timePart := make([]byte, 10)
	for i := 9; i >= 0; i-- {
		timePart[i] = crockford[ms%32]
		ms /= 32
	}
	entropy := make([]byte, 10)
	if _, err := rand.Read(entropy); err != nil {
		return "", err
	}
	return string(timePart) + crockfordEncoding.EncodeToString(entropy), nil
}

func LocalISO(t time.Time) string {
	return t.Format("2006-01-02T15:04:05-07:00")
}

func Serialize(n Record) (string, error) {
	out, err := jsonx.Compact(n)
	if err != nil {
		return "", err
	}
	return string(out), nil
}

var dateShapePattern = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}T.*(?:Z|[+-]\d{2}:\d{2})$`)

var noteTimeLayouts = []string{
	time.RFC3339Nano,
	time.RFC3339,
	"2006-01-02T15:04Z07:00",
}

func ParseDate(s string) (time.Time, bool) {
	for _, layout := range noteTimeLayouts {
		if t, err := time.Parse(layout, s); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

func IsShaped(n Record) bool {
	if n.ID == "" || n.Date == "" || n.Body == "" {
		return false
	}
	if !slices.Contains(Types, n.Type) || !slices.Contains(Authors, n.Author) {
		return false
	}
	if !dateShapePattern.MatchString(n.Date) {
		return false
	}
	if len(n.Date) < 10 || !models.IsValidYMD(n.Date[:10]) {
		return false
	}
	if _, ok := ParseDate(n.Date); !ok {
		return false
	}
	return true
}

func Parse(raw []byte) (Record, bool) {
	var n Record
	if err := json.Unmarshal(raw, &n); err != nil {
		return Record{}, false
	}
	if !IsShaped(n) {
		return Record{}, false
	}
	return n, true
}

func Compare(a, b Record) int {
	at, _ := ParseDate(a.Date)
	bt, _ := ParseDate(b.Date)
	if !at.Equal(bt) {
		if at.Before(bt) {
			return -1
		}
		return 1
	}
	return strings.Compare(a.ID, b.ID)
}

func Write(p paths.DataPaths, n Record) error {
	if err := os.MkdirAll(p.NotesDir, 0o755); err != nil {
		return err
	}
	serialized, err := Serialize(n)
	if err != nil {
		return err
	}
	return atomicfile.Write(filepath.Join(p.NotesDir, n.ID+".json"), []byte(serialized+"\n"), 0o644)
}

type looseEntry struct {
	note Record
	file string
}

func readLooseEntries(p paths.DataPaths) ([]looseEntry, error) {
	files, err := os.ReadDir(p.NotesDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	names := make([]string, 0, len(files))
	for _, f := range files {
		if strings.HasSuffix(f.Name(), ".json") {
			names = append(names, f.Name())
		}
	}
	sort.Strings(names)

	var entries []looseEntry
	for _, name := range names {
		data, err := os.ReadFile(filepath.Join(p.NotesDir, name))
		if err != nil {
			return nil, err
		}
		if n, ok := Parse(data); ok {
			entries = append(entries, looseEntry{note: n, file: name})
		}
	}
	return entries, nil
}

func readArchived(p paths.DataPaths) (map[string]Record, error) {
	out := map[string]Record{}
	files, err := os.ReadDir(p.ArchiveDir)
	if err != nil {
		if os.IsNotExist(err) {
			return out, nil
		}
		return nil, err
	}
	for _, f := range files {
		if !strings.HasSuffix(f.Name(), ".jsonl") {
			continue
		}
		data, err := os.ReadFile(filepath.Join(p.ArchiveDir, f.Name()))
		if err != nil {
			return nil, err
		}
		for line := range bytes.SplitSeq(data, []byte("\n")) {
			if len(bytes.TrimSpace(line)) == 0 {
				continue
			}
			if n, ok := Parse(line); ok {
				if _, exists := out[n.ID]; !exists {
					out[n.ID] = n
				}
			}
		}
	}
	return out, nil
}

func ReadAll(p paths.DataPaths) ([]Record, error) {
	merged, err := readArchived(p)
	if err != nil {
		return nil, err
	}
	loose, err := readLooseEntries(p)
	if err != nil {
		return nil, err
	}
	for _, e := range loose {
		merged[e.note.ID] = e.note
	}
	out := make([]Record, 0, len(merged))
	for _, n := range merged {
		out = append(out, n)
	}
	slices.SortFunc(out, Compare)
	return out, nil
}

type Filter struct {
	Type      string
	Since     string
	WorkoutID *int64
}

func Apply(records []Record, f Filter) []Record {
	out := make([]Record, 0, len(records))
	for _, n := range records {
		if f.Type != "" && n.Type != f.Type {
			continue
		}
		if f.Since != "" && n.Date[:10] < f.Since {
			continue
		}
		if f.WorkoutID != nil && (n.WorkoutID == nil || *n.WorkoutID != *f.WorkoutID) {
			continue
		}
		out = append(out, n)
	}
	return out
}

type archiveYear struct {
	notes         []Record
	safeToRewrite bool
}

func readArchiveYear(p paths.DataPaths, year int) archiveYear {
	data, err := os.ReadFile(p.ArchiveFile(year))
	if err != nil {
		return archiveYear{safeToRewrite: os.IsNotExist(err)}
	}
	seen := map[string]bool{}
	var out []Record
	for line := range bytes.SplitSeq(data, []byte("\n")) {
		if len(bytes.TrimSpace(line)) == 0 {
			continue
		}
		n, ok := Parse(line)
		if !ok {
			return archiveYear{safeToRewrite: false}
		}
		if seen[n.ID] {
			return archiveYear{safeToRewrite: false}
		}
		seen[n.ID] = true
		out = append(out, n)
	}
	return archiveYear{notes: out, safeToRewrite: true}
}

type CompactResult struct {
	Archived     int
	Years        []int
	SkippedYears []int
}

func Compact(p paths.DataPaths, now time.Time) (CompactResult, error) {
	cutoff := now.AddDate(0, 0, -compactAgeDays)

	entries, err := readLooseEntries(p)
	if err != nil {
		return CompactResult{}, err
	}
	deduped := map[string]Record{}
	contentByID := map[string]string{}
	divergent := map[string]bool{}
	for _, e := range entries {
		content, err := Serialize(e.note)
		if err != nil {
			return CompactResult{}, err
		}
		if prior, ok := contentByID[e.note.ID]; ok && prior != content {
			divergent[e.note.ID] = true
		}
		contentByID[e.note.ID] = content
		deduped[e.note.ID] = e.note
	}

	var eligible []Record
	for _, n := range deduped {
		if divergent[n.ID] {
			continue
		}
		t, ok := ParseDate(n.Date)
		if !ok || !t.Before(cutoff) {
			continue
		}
		eligible = append(eligible, n)
	}
	if len(eligible) == 0 {
		return CompactResult{}, nil
	}

	byYear := map[int][]Record{}
	for _, n := range eligible {
		year, err := strconv.Atoi(n.Date[:4])
		if err != nil {
			continue
		}
		byYear[year] = append(byYear[year], n)
	}

	if err := os.MkdirAll(p.ArchiveDir, 0o755); err != nil {
		return CompactResult{}, err
	}

	result := CompactResult{}
	years := make([]int, 0, len(byYear))
	for year := range byYear {
		years = append(years, year)
	}
	sort.Ints(years)

	for _, year := range years {
		batch := byYear[year]
		existing := readArchiveYear(p, year)
		if !existing.safeToRewrite {
			result.SkippedYears = append(result.SkippedYears, year)
			continue
		}
		merged := map[string]Record{}
		for _, n := range existing.notes {
			merged[n.ID] = n
		}
		for _, n := range batch {
			merged[n.ID] = n
		}
		combined := make([]Record, 0, len(merged))
		for _, n := range merged {
			combined = append(combined, n)
		}
		slices.SortFunc(combined, Compare)

		var buf bytes.Buffer
		for _, n := range combined {
			serialized, err := Serialize(n)
			if err != nil {
				return CompactResult{}, err
			}
			buf.WriteString(serialized)
			buf.WriteByte('\n')
		}
		if err := atomicfile.Write(p.ArchiveFile(year), buf.Bytes(), 0o644); err != nil {
			result.SkippedYears = append(result.SkippedYears, year)
			continue
		}

		archivedIDs := map[string]bool{}
		for _, n := range batch {
			archivedIDs[n.ID] = true
		}
		for _, e := range entries {
			if archivedIDs[e.note.ID] {
				os.Remove(filepath.Join(p.NotesDir, e.file))
			}
		}
		result.Archived += len(batch)
		result.Years = append(result.Years, year)
	}

	sort.Ints(result.Years)
	sort.Ints(result.SkippedYears)
	return result, nil
}
